packages/java-edition/src/json/checker/index.ts

```diff
@@ -82,20 +82,73 @@ export function definition(
 	identifier: `::${string}::${string}`,
 ): core.SyncChecker<JsonNode> {
 	return (node, ctx) => {
-		const symbol = ctx.symbols.query(ctx.doc, 'mcdoc', identifier)
-		const typeDef = symbol.getData(mcdoc.binder.TypeDefSymbolData.is)?.typeDef
-		if (!typeDef) {
-			return
-		}
-		switch (typeDef.kind) {
-			case 'struct':
-				object(typeDef)(node, ctx)
-				break
-			default:
-				ctx.logger.error(
-					`[json.checker.definition] Expected a struct type, but got ${typeDef.kind}`,
-				)
-		}
+		mcdoc.validator.reference<JsonNode | core.PairNode<JsonStringNode, JsonNode>>(node, identifier, {
+			context: ctx,
+			inferType: node => {
+				switch (node.type) {
+					case 'json:boolean': return { kind: 'literal', value: { kind: 'boolean', value: node.value! } }
+					case 'json:number': return { kind: 'literal', value: { kind: node.value.type, value: Number(node.value.value) } }
+					case 'json:null': return { kind: 'union', members: [] } // null is always invalid? 
+					case 'json:string': return { kind: 'literal', value: { kind: 'string', value: node.value } }
+					case 'json:array': return { kind: 'list', item: { kind: 'any' } }
+					case 'json:object': return { kind: 'struct', fields: [] }
+					case 'pair': 
+						if (!node.key) {
+							return { kind: 'tuple', items: [] };
+						}
+						return { kind: 'tuple', items: [ { kind: 'literal', value: { kind: 'string', value: node.key.value }}, { kind: 'any' } ] }
+				}
+			},
+			isEquivalent: (inferred, def) => {
+				switch (inferred.kind) {
+					case 'list':
+						return (['list', 'byte_array', 'int_array', 'long_array', 'tuple'] as mcdoc.McdocType['kind'][]).includes(def.kind);
+					case 'struct':
+						return def.kind === 'struct';
+					case 'byte':
+					case 'short':
+					case 'int':
+					case 'long':
+						return [ 'byte', 'short', 'int', 'long', 'float', 'double' ].includes(def.kind);
+					case 'float':
+					case 'double':
+						return [ 'float', 'double' ].includes(def.kind);
+					default: return false;
+				}
+			},
+			getChildren: node => {
+				if (node.type === 'pair') {
+					return [ node.key, node.value ].filter(n => n) as JsonNode[]
+				}
+				if (node.type === 'json:array') {
+					return node.children.filter(n => n.value).map(n => n.value as JsonNode)
+				}
+				return node.children?.filter(n => n) as JsonNode[] ?? []
+			},
+			getRange: (node, err) => {
+				if ((node.type === 'json:object' && err === 'missing_key')
+					|| node.type === 'json:array' && err === 'invalid_collection_length') {
+					return { start: node.range.start, end: node.range.start }
+				}
+				return node.range;
+			},
+			attachTypeInfo: (node, definition) => {}, //TODO
+		});
+
```

Comment:
I assume this would put the mcdoc information on the json nodes to be used in the completer for example?

Reply:
Yes

Reply:
I'm not sure this info is enough btw, binder probably needs the original symbol and further path somehow?